### ISSUE TITLE
Implement CARMA in Python

### ISSUE THREAD

**Issue by ireneisdoomed** [opentargets/issues#3105]
We have been working in a Python version of the CARMA finemapping package, originally written in R.
The Python implementation would allow us:
- to have ownership on the logic that produces PIPs
- to have compatibility with our new OTG framework

## Context
Original package: https://github.com/ZikunY/CARMA
Most of the logic is written in native R in a very ad hoc manner, often dealing with nested structures and not dataframes.
The algorithm that computes the **marginal likelihoods** is written in C++.

### CARMA inputs
- Input data:
    1.  Z scores for each SNP extracted from summary statistics.
        -  **Structure**. Array of one element: list of Z scores (unordered by value or SNP position)
    2. LD correlation matrix.
        - **Structure**. Array of one element: square matrix with r values. Variant IDs as column and row names.

<details open>
<summary>Other parameters:</summary>
<br>

| Parameter                     | Description                                                                                                              | Default                | Notes                                                                                                                  |
|-------------------------------|--------------------------------------------------------------------------------------------------------------------------|------------------------|------------------------------------------------------------------------------------------------------------------------|
| _w_list_                      | functional annotations matrix at each individual locus.                                                                  | Null                   |                                                                                                                        |
| _lambda_list_                 | hyper-parameter at each individual locus.                                                                                | Null                   |                                                                                                                        |
| _label_list_                  | list of the names at the testing loci                                                                                    | Null                   |                                                                                                                        |
| _effect_size_prior_           | Assumed model for the prior distribution of effect sizes.                                                                | Spike-slab             |                                                                                                                        |
| _tau_                         | Parameter used to set the effect size priors                                                                             | 0.04                   |                                                                                                                        |
| _input_alpha_                 | Elastic net mixing parameter. I think this has to do with the penalisation function                                      | 0                      |                                                                                                                        |
| _epsilon_threshold_           | Convergence threshold measured by average of Bayes factors                                                               | 1e-4                  |                                                                                                                        |
| _y_var_                       | input variance of the summary statistics                                                                                 | 1                      | as the summary statistics are standardised                                                                              |
| _rho_index_                   | parameter in the credible set estimation.                                                                                | 0.99                   |                                                                                                                        |
| _BF_index_                    | threshold of the Bayes factor for credible model estimation.                                                             | 10                     | documentation says that it has to be "a number smaller than 1", I wonder if it should be 0.1                           |
| _outlier_BF_index_            | threshold of the Bayes factor for outlier detection                                                                      | 1/3.2                 |                                                                                                                        |
| _EM_dist_                     | distribution used to model the prior probability of being causal as a function of functional annotations.                 | Logistic               |                                                                                                                        |
| _prior_prob_computation_      | Distribution for the penalisation function                                                                              | Logistic               |                                                                                                                        |
| _num_causal_                  | maximum number of causal variants assumed per locus                                                                      | 10                     | this means that we'll have 10 credible sets                                                                             |
| _Max_Model_Dim_               | Maximum number of the top candidate models based on the unnormalized posterior probability                               | 2e5                    |                                                                                                                        |
| _all_iter_                    | Maximum iterations for expectation-maximization algorithm to run.                                                        | 3                      |                                                                                                                        |
| _all_inner_iter_              | Maximum iterations for Shotgun algorithm to run per iteration                                                            | 10                     | I think this needs to equal the maximum number of causal variants                                                       |
</details>

### CARMA outputs
The result of running `carma_fixed_sigma` has a structure as follows. It is one array of one element, that is another list containing 4 elements:
_Based on my test locus with 481 SNPs_

| Output            | Description                                                                                             | Structure                                                                                                                                                               | Calculated by                              |
|-------------------|---------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------|
| PIPs              | Posterior probability per locus, this is the main result of the Cauchy algorithm.                       | Array of n elements. As many as SNPs in the locus.                                                                                                                      | `Module.Cauchy.Shotgun`                   |
| Credible set      | Technically contains credible set containing the most likely causal variants. I don't understand the output.  | List of 2 elements: 1.  Array of 3 elements; 2. Array of 9 elements.                                                                                                    | `Credible.set.fun.improved`               |
| Credible model    | Technically contains credible models that meet the specified Bayes threshold. I don't understand the output.  | List of 4 elements: 1. Array of 4 vectors of size 11, storing SNP indices. 2. Dataframe of 4 PIPs; 3. Vector of 15 SNP indices                                         | `Credible.Model.Fun`                      |
| Outliers          | List of SNPs identified as outliers.                                                                     | List of 2 elements: 1. Array of SNP indices identified as outliers; 2. Their Z scores                                                                                  | The C++ modules. In our case, `Outlier_ind_normal_marginal` |

## CARMA components
1. Calculation of posterior probabilities to define credible sets and models accounting for multiple causal variants in a single locus.
- Implementation of the Shotgun algorithm.
2. Detection of outlier SNPs that show discrepancies between their LD and Z scores.
- This algorithm is embedded in the Shotgun method. This means that we cannot detect outliers without having an accurate calculation of PIPs.
- The manuscript explains that outliers will lead to aberrant, high PIPs.
- Their proposed method calculates a regularised LD matrix (applying Ridge correction), and makes a contrast hypothesis test calculating PIPs using the normalised and the original version ([see Supp Methods](https://www.nature.com/articles/s41588-023-01392-0)).

**Comment by ireneisdoomed** [comment]
## Python Implementation Issues with Carmapy 0.1.0
### Overview
Author: @hlnicholls
Code Repository: https://github.com/hlnicholls/carmapy/tree/0.1.0
The goal was to replicate the R logic using Python, particularly leveraging Numpy for matrix operations.

### Issue
The Python version fails to produce accurate results:
- The code crashes.
- Intermediate results deviate from expectations.

### Observations & Attempts

**1. Testing & Debugging:**
- I created a [Colab notebook](https://colab.research.google.com/drive/18WPgox3xkH3lAcNE34RD8H4KGBtHuylh?usp=sharing) for testing using the original R function and APOE locus test data found in the [tests](https://github.com/hlnicholls/carmapy/raw/main/tests) directory.
- The Python version was refactored to separate the functions (originally nested) to simplify debugging.

**2. Inputs & Matrix Creation:**
- The main function receives lists of Z and LD scores.
- The Python and R implementations differ in matrix creation.
- In Python, the Z score matrix was transposed, affecting all subsequent operations.
- I've ensured inputs to functions maintain the expected structure.

**3. Issues in `set_gamma_func`:**
- This crucial function wasn't performing as anticipated, necessitating a complete rewrite.
- Expected output: 2 elements. Observed output: 3 elements.
- Discrepancies were noticed between Python and R in the transformation of the submatrix (the second element).

### Suggestions
In my opinion, we should:
- Avoid direct line-by-line translation from R to Python. The current R implementation is inefficient and can be improved. For example, it constantly iterates over arrays that are actually of size 1.
- Breaking the R code into separate functions would make refactoring much easier.
- Simplify the output structure. We have to determine the essential outputs and split them in different fields.

Given that `set_gamma_func` needed a rewrite, I expect that other functions might as well. Properly addressing these issues will require considerable time and resources.

**Comment by addramir** [comment]
While we are working on MVP for the pipline, I have developed the simplified light version of the CARMA:
https://github.com/addramir/carmapy_light

**What changed:**
1. It uses only Spike-slab effect size priors and Poisson model priors.
2. C++ is re-implemented in Python.
3. The way of storing the configuration list is changed. It uses a string with the list of indexes for causal SNPs instead of a sparse matrix.
4. Fixed bugs in PIP calculation.
5. No credible models.
6. No credible sets, only PIPs.
7. No functional annotations.
8. Removed unnecessary parameters.
